### PR TITLE
feat(topology): add new service in context

### PR DESCRIPTION
**Pull request by nemesis09**
This PR-
* changes the connector icon when not over a node
* adds a context menu on releasing connector on graph or application group
* adds add-resources menu as the context menu on release of connector on graph or application group
* adds necessary URL params to context menu options for creating connection after save
* is tracked by issue https://issues.redhat.com/browse/ODC-2352

Screen (demo implementation on top of #3918 (add-resources menu) with changes pulled from #3958 (creating the connection post save) to show complete flow) -
in-context-add

### CODE REVIEW

**Comment by andrewballantyne** [comment]
@nemesis09 Could you post a link to your ticket on your PR even if it's a WIP. 

/kind feature

**Comment by nemesis09** [comment]
putting on hold as it needs PR https://github.com/openshift/console/pull/3918 to be merged first.
/hold

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 1]
```diff
@@ -226,8 +254,42 @@ const edgeDragSourceSpec = (
...
+    }
+  };
+
+  if (isGraph(target)) {
+    if (!choice) {
+      return graphActions(target.getController().getElements());
+    }
+    onKebabOptionClick(choice, source);
+    return graphActions(target.getController().getElements());
+  }
```
Comment (karthikjeeyar) [inline frontend/packages/dev-console/src/components/topology/componentUtils.ts]: you could simplify it as  
```if (isGraph(target)) {
    if (choice) {
      onKebabOptionClick(choice, source);
    }
    return graphActions(target.getController().getElements());
  }

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 2]
```diff
@@ -226,8 +254,42 @@ const edgeDragSourceSpec = (
...
+    const applicationGroup: TopologyApplicationObject = {
+      id: target.getId(),
+      name: target.getLabel(),
+      resources: target.getNodes().map((currNode) => currNode.getData()),
+    };
+    if (!choice) {
+      return groupActions(applicationGroup, true);
+    }
+    onKebabOptionClick(choice, source);
+    return groupActions(applicationGroup, true);
```
Comment (karthikjeeyar) [inline frontend/packages/dev-console/src/components/topology/componentUtils.ts]: same as above
   ```
    if (choice) {
      onKebabOptionClick(choice, source);
    }
    return groupActions(applicationGroup, true);

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 1 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: Done. PTAL

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 2 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: Done. PTAL

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx [hunk 3]
```diff
@@ -156,20 +157,21 @@ const CreateConnectorWidget: React.FC<CreateConnectorWidgetProps> = observer((pr
       {prompt && (
         <ContextMenu
           reference={{ x: prompt.event.pageX, y: prompt.event.pageY }}
+          className="odc-topology-context-menu"
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: I've passed it as a prop to withCreateConnector()
PTAL

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 4]
```diff
@@ -226,8 +254,40 @@ const edgeDragSourceSpec = (
...
+  choice,
 ): any[] | null => {
+  const onKebabOptionClick = (option: KebabOption, sourceNode: Node) => {
+    if (option.callback) {
+      option.callback();
+    }
+    if (option.href) {
+      const sourceObj = _.get(sourceNode.getData(), ['resources', 'obj'], null);
+      const hrefWithContext = `${option.href}&contextSource=${referenceFor(sourceObj)}/${
+        sourceObj.metadata.name
```
Comment (karthikjeeyar) [inline frontend/packages/dev-console/src/components/topology/componentUtils.ts]: we should not be able to create knative resources using the drag connector, because knative revision can only be a source, it can never be a target. add `isKnativeDisabled` param here and remove it from other place.

Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: added `isKnativeDisabled` param

**Comment by nemesis09** [comment]
removing hold as dependent PR https://github.com/openshift/console/pull/3918 is merged
/hold cancel

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 4 above)
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: We shouldn't need this param. The add forms should properly handle this case by identifying there is a `contextSource` present and therefore remove knative service option.

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 5]
```diff
@@ -226,8 +254,40 @@ const edgeDragSourceSpec = (
...
+      onKebabOptionClick(choice, source);
+    }
+    return graphActions(target.getController().getElements());
+  }
+  if (target.isGroup()) {
+    const applicationGroup: TopologyApplicationObject = {
+      id: target.getId(),
+      name: target.getLabel(),
+      resources: target.getNodes().map((currNode) => currNode.getData()),
+    };
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/componentUtils.ts]: Why is this here? i see it's copied code. @jeff-phillips-18 why do we need this? Shouldn't we be using the node data directly?

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx [hunk 6]
```diff
@@ -162,20 +165,21 @@ const CreateConnectorWidget: React.FC<CreateConnectorWidgetProps> = observer((pr
...
-            >
-              {c.label}
+            <ContextMenuItem key={c.label}>
+              {
+                <KebabItem
+                  option={c}
+                  onClick={() => onCreate(prompt.element, prompt.target, prompt.event, c)}
+                />
+              }
             </ContextMenuItem>
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/withCreateConnector.tsx]: We cannot change this to use a `KebabItem` because this is generic code that shouldn't depend on console.
We'll need to find a way to support the console use case without this dependency.
Doesn't this break the storybook sample?
If it doesn't break it then I suppose we can keep it with a TODO to fix later.
But otherwise we need a way to supply a list of react components that go into the context menu.

Maybe we simply add the ability to supply a react element as a choice instead:

```
choices: ConnectorChoice[] | React.ReactNode;
```

```
         {
           React.isValidElement(prompt.choices) ? prompt.choices : 
            prompt.choices.map((c) => (
            <ContextMenuItem
              key={c.label}
              onClick={() => {
                onCreate(prompt.element, prompt.target, prompt.event, c);
              }}
            >
              {c.label}
            </ContextMenuItem>
          ))}
```

Then in your `onCreate` function instead of returning the list of actions, you actually build the menu items like we already do elsewhere. You'd be able to use the kebab actions handling of the callback / href without duplicating the code.

File: frontend/packages/dev-console/src/components/topology/componentFactory.ts [hunk 7]
```diff
@@ -194,6 +197,30 @@ class ComponentFactory {
...
+                        withContextMenu(
+                          workloadContextMenu,
+                          document.getElementById('modal-container'),
+                          'odc-topology-context-menu',
+                        )(WorkloadNode),
+                      ),
+                    ),
+                  ),
+                ),
+              );
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/componentFactory.ts]: Why is this here? We should already have explicit support for each known type already.

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 8]
```diff
@@ -152,13 +158,25 @@ const graphWorkloadDropTargetSpec: DropTargetSpec<
...
 > = {
-  accept: [TYPE_WORKLOAD, TYPE_KNATIVE_SERVICE, TYPE_EVENT_SOURCE, TYPE_CONNECTS_TO],
+  accept: [
+    TYPE_WORKLOAD,
+    TYPE_KNATIVE_SERVICE,
+    TYPE_EVENT_SOURCE,
+    TYPE_CONNECTS_TO,
+    MOVE_CONNECTOR_DROP_TYPE,
+    CREATE_CONNECTOR_DROP_TYPE,
+    CREATE_CONNECTOR_OPERATION,
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: done

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 9]
```diff
@@ -152,13 +158,25 @@ const graphWorkloadDropTargetSpec: DropTargetSpec<
   { dragEditInProgress: boolean },
   GraphProps
 > = {
-  accept: [TYPE_WORKLOAD, TYPE_KNATIVE_SERVICE, TYPE_EVENT_SOURCE, TYPE_CONNECTS_TO],
+  accept: [
+    TYPE_WORKLOAD,
+    TYPE_KNATIVE_SERVICE,
+    TYPE_EVENT_SOURCE,
+    TYPE_CONNECTS_TO,
+    MOVE_CONNECTOR_DROP_TYPE,
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: done

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 10]
```diff
@@ -152,13 +158,25 @@ const graphWorkloadDropTargetSpec: DropTargetSpec<
...
+    CREATE_CONNECTOR_DROP_TYPE,
+    CREATE_CONNECTOR_OPERATION,
+  ],
   canDrop: (item, monitor, props) => {
-    return monitor.getOperation() === REGROUP_OPERATION && item.getParent() !== props.element;
+    return (
+      (monitor.getOperation() === REGROUP_OPERATION && item.getParent() !== props.element) ||
+      monitor.getOperation() === CREATE_CONNECTOR_OPERATION
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: done

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 11]
```diff
@@ -167,13 +185,23 @@ const groupWorkloadDropTargetSpec: DropTargetSpec<
...
+    TYPE_WORKLOAD,
+    TYPE_EVENT_SOURCE,
+    TYPE_KNATIVE_SERVICE,
+    MOVE_CONNECTOR_DROP_TYPE,
+    CREATE_CONNECTOR_DROP_TYPE,
+    CREATE_CONNECTOR_OPERATION,
+  ],
+  canDrop: (item, monitor) =>
+    monitor.getOperation() === REGROUP_OPERATION ||
+    monitor.getOperation() === CREATE_CONNECTOR_OPERATION,
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: done

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 12]
```diff
@@ -167,13 +185,23 @@ const groupWorkloadDropTargetSpec: DropTargetSpec<
...
 > = {
-  accept: [TYPE_WORKLOAD, TYPE_EVENT_SOURCE, TYPE_KNATIVE_SERVICE],
-  canDrop: (item, monitor) => monitor.getOperation() === REGROUP_OPERATION,
+  accept: [
+    TYPE_WORKLOAD,
+    TYPE_EVENT_SOURCE,
+    TYPE_KNATIVE_SERVICE,
+    MOVE_CONNECTOR_DROP_TYPE,
+    CREATE_CONNECTOR_DROP_TYPE,
+    CREATE_CONNECTOR_OPERATION,
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: done

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 13]
```diff
@@ -167,13 +185,23 @@ const groupWorkloadDropTargetSpec: DropTargetSpec<
   { droppable: boolean; dropTarget: boolean; canDrop: boolean },
   any
 > = {
-  accept: [TYPE_WORKLOAD, TYPE_EVENT_SOURCE, TYPE_KNATIVE_SERVICE],
-  canDrop: (item, monitor) => monitor.getOperation() === REGROUP_OPERATION,
+  accept: [
+    TYPE_WORKLOAD,
+    TYPE_EVENT_SOURCE,
+    TYPE_KNATIVE_SERVICE,
+    MOVE_CONNECTOR_DROP_TYPE,
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: done

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 14]
```diff
@@ -226,8 +254,40 @@ const edgeDragSourceSpec = (
...
+  }
+  if (target.isGroup()) {
+    const applicationGroup: TopologyApplicationObject = {
+      id: target.getId(),
+      name: target.getLabel(),
+      resources: target.getNodes().map((currNode) => currNode.getData()),
+    };
+    if (choice) {
+      onKebabOptionClick(choice, source);
+    }
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: done

File: frontend/packages/dev-console/src/components/topology/componentFactory.ts (hunk 7 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentFactory.ts]: It was overlooked while cleanup.
removed it.

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx (hunk 6 above)
Comment (karthikjeeyar) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: @nemesis09  Check should be on item, `React.isValidElement(prompt.choices?.[0]`

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx [hunk 15]
```diff
@@ -1,5 +1,6 @@
 import * as React from 'react';
 import { observer } from 'mobx-react';
+import { KebabItem } from '@console/internal/components/utils';
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: Removed KebabItem from topology

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 5 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: we could use node data directly but that would require modifying the groupContextMenu and groupActions files. so I created this object here.
I've made the necessary modifications to the functions and groupContextMenu is directly used now.

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx (hunk 6 above)
Comment (nemesis09) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: Done

File: frontend/packages/dev-console/src/components/topology/componentFactory.ts (hunk 7 above)
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/topology/componentFactory.ts]: @nemesis09  please remove this block.

**Comment by karthikjeeyar** [comment]
While dropping the connector the whole page flickers, @nemesis09 please confirm if there needs to be a followup bug for this.


AwesomeScreenshot-2020-1-24-1579854490247

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 16]
```diff
@@ -226,8 +243,27 @@ const edgeDragSourceSpec = (
...
   source: Node,
-  target: Node,
+  target: Node | Graph,
 ): any[] | null => {
+  const onKebabOptionClick = (option: KebabOption) => {
+    if (option.callback) {
+      option.callback();
+    }
+    if (option.href) {
+      const sourceObj = _.get(source.getData(), ['resources', 'obj'], null);
```
Comment (divyanshiGupta) [inline frontend/packages/dev-console/src/components/topology/componentUtils.ts]: You can use optional chaining here.

**Comment by invincibleJai** [comment]
@nemesis09 will there be a + icon for a connector within a group, currently it's arrow

Screenshot 2020-01-24 at 5 35 41 PM

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 16 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: should we allow creation of resources through this flow if sourceObj is null or undefined. since it will reduce the flow to a simple add-resource flow?

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx [hunk 17]
```diff
@@ -23,9 +23,14 @@ const onKebabOptionClick = (option: KebabOption) => {
...
-const createMenuItems = (actions: KebabMenuOption[]) =>
-  actions.map((option) =>
-    isKebabSubMenu(option) ? (
+const createMenuItems = (actions: KebabMenuOption[], onOptionClick?: (item) => void) => {
+  return actions.map((option) =>
+    onOptionClick ? (
+      <ContextMenuItem
+        key={option.label}
+        component={<KebabItem option={option} onClick={() => onOptionClick(option)} />}
+      />
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: done

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx [hunk 18]
```diff
@@ -23,34 +23,47 @@ const onKebabOptionClick = (option: KebabOption) => {
   }
 };
 
-const createMenuItems = (actions: KebabMenuOption[]) =>
-  actions.map((option) =>
+const createMenuItems = (actions: KebabMenuOption[], onOptionClick?: (item) => void) => {
```
Comment (karthikjeeyar) [inline frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]:
```suggestion
const createMenuItems = (actions: KebabMenuOption[], onOptionClick?: (item: KebabOption) => void) => {
```

**Comment by nemesis09** [comment]
> While dropping the connector the whole page flickers, @nemesis09 please confirm if there needs to be a followup bug for this.
> 
> AwesomeScreenshot-2020-1-24-1579854490247
> cc: @christianvogt

I noticed the flicker too, but it only occurs in some of the cases when the connector is dropped towards the boundaries of the window. I think we can track this in a follow up bug.

File: frontend/packages/dev-console/src/components/topology/componentFactory.ts (hunk 7 above)
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/topology/componentFactory.ts]: @nemesis09  Please remove this code

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx (hunk 18 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: done

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 16 above)
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: In which case the sourceObj will be undefined though? Dragging sourceObj connector and on releasing we will see the menu to create the service.

Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: the sourceObj 

> @nemesis09 will there be a + icon for a connector within a group, currently it's arrow
> 
> Screenshot 2020-01-24 at 5 35 41 PM

Done

**Comment by nemesis09** [comment]
> @nemesis09 will there be a + icon for a connector within a group, currently it's arrow
> 
> Screenshot 2020-01-24 at 5 35 41 PM

Fixed this.

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 16 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: The only time i saw an error was when component did not yet mount. other than that, I think the source will always be there.

**Comment by andrewballantyne** [comment]
> > While dropping the connector the whole page flickers, @nemesis09 please confirm if there needs to be a followup bug for this.
> > AwesomeScreenshot-2020-1-24-1579854490247
> > cc: @christianvogt
> 
> I noticed the flicker too, but it only occurs in some of the cases when the connector is dropped towards the boundaries of the window. I think we can track this in a follow up bug.

I do not see this... how did you guys get this to happen?

**Comment by invincibleJai** [comment]
@andrewballantyne drag the connector and drop immediately if we drag and hold for a sec before dropping i don't see as well (am on mac)

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 16 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: updated the code with suggested change

**Comment by karthikjeeyar** [comment]
Verified locally
/lgtm

**Comment by andrewballantyne** [comment]
> @andrewballantyne drag the connector and drop immediately if we drag and hold for a sec before dropping i don't see as well (am on mac)

Interesting. I'll try to reproduce this. I moved pretty quick I though. 🤷‍♂ Is it a 100% for you guys?

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 16 above)
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: If you have seen the error probably do not allow redirect to create flow if sourceObj is undefined

**Comment by nemesis09** [comment]
> > @andrewballantyne drag the connector and drop immediately if we drag and hold for a sec before dropping i don't see as well (am on mac)
> 
> Interesting. I'll try to reproduce this. I moved pretty quick I though. man_shrugging Is it a 100% for you guys?

no. It is not always reproducible for me.

**Comment by karthikjeeyar** [comment]
> > @andrewballantyne drag the connector and drop immediately if we drag and hold for a sec before dropping i don't see as well (am on mac)
> 
> Interesting. I'll try to reproduce this. I moved pretty quick I though.  Is it a 100% for you guys?

@andrewballantyne Move a node below the lower half of the screen and drag the connector and drop it

File: frontend/packages/topology/src/components/DefaultCreateConnector.tsx [hunk 19]
```diff
@@ -23,7 +23,7 @@ const DefaultCreateConnector: React.FC<DefaultCreateConnectorProps> = ({
       x2={endPoint.x}
       y2={endPoint.y}
     />
-    {hints && hints.length === 1 && hints[0] === 'create' ? (
+    {hints && hints[hints.length - 1] === 'create' ? (
```
Comment (nemesis09) [inline frontend/packages/topology/src/components/DefaultCreateConnector.tsx]: modified the check to allow the latest hint to decide the connector type
cc @christianvogt

**Comment by andrewballantyne** [comment]
connectorChangesHandle

It's a little hard to see, but the "+" arrow-head changes back and forth between the triangle arrow-head. @christianvogt you built the original conversion... did this happen in your example (ie is this some sort of implementation issue or a bug with this work)?

Also, still can't reproduce the page flicker. Is a number of nodes? The only "flicker" I get is the highlight disappears a frame before the menu pops up, so my eyes jut around the screen looking at changes.

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx [hunk 20]
```diff
@@ -162,22 +164,25 @@ const CreateConnectorWidget: React.FC<CreateConnectorWidgetProps> = observer((pr
...
-            <ContextMenuItem
-              key={c.label}
-              onClick={() => {
-                onCreate(prompt.element, prompt.target, prompt.event, c);
-              }}
-            >
-              {c.label}
-            </ContextMenuItem>
-          ))}
+          {React.isValidElement(prompt.choices?.[0])
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/withCreateConnector.tsx]:
```suggestion
          {React.isValidElement(prompt.choices)
```

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx [hunk 21]
```diff
@@ -30,6 +30,7 @@ type CreateConnectorWidgetProps = {
     choice?: ConnectorChoice,
   ) => ConnectorChoice[] | void | undefined | null;
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/withCreateConnector.tsx]:
```suggestion
  ) => ConnectorChoice[] | void | undefined | null | React.ReactElement[];
```

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 22]
```diff
@@ -226,8 +242,27 @@ const edgeDragSourceSpec = (
...
+      option.callback();
+    }
+    if (option.href) {
+      const sourceObj = source.getData()?.resources?.obj || {};
+      const hrefWithContext = `${option.href}&contextSource=${referenceFor(sourceObj)}/${
+        sourceObj?.metadata?.name
+      }`;
+      history.push(hrefWithContext);
+    }
+  };
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/componentUtils.ts]: We shouldn't have to override and duplicate behavior. Parameterize our actions like @karthikjeeyar did in `add-resources-menu-utils` to support `addApplication`. We should also support a context source object. This would put all the logic for creation of these actions in a single place.
The resulting code would be inside `#getAddPageUrl` to create the href.

I also do not think we need `isKnativeDisabled` as a query param in the URL. By @karthikjeeyar logic for the need, the fact that we'll have a `contextSource` will give us enough info to determine that we can hide the knative service option.

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 23]
```diff
@@ -226,8 +242,27 @@ const edgeDragSourceSpec = (
 
 const createConnectorCallback = (serviceBinding: boolean) => (
   source: Node,
-  target: Node,
+  target: Node | Graph,
 ): any[] | null => {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/componentUtils.ts]:
```suggestion
): React.ReactElement[] | null => {
```

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx [hunk 24]
```diff
@@ -23,34 +23,56 @@ const onKebabOptionClick = (option: KebabOption) => {
...
+    : createMenuItems(kebabOptionsToMenu(groupActions(applicationData)));
 };
 export const nodeContextMenu = (element: Node) =>
   createMenuItems(kebabOptionsToMenu(nodeActions(element.getData())));
 
-export const graphContextMenu = (element: Graph) =>
-  createMenuItems(kebabOptionsToMenu(graphActions(element.getController().getElements())));
+export const graphContextMenu = (
+  element: Graph,
+  connectorDropMenuCallback?: (item: KebabOption) => void,
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: Lots of changes will need to be made to this file to support my suggestion about improving the href generation in the actions utils.

```suggestion
  contextSource?: <type>,
```

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx [hunk 25]
```diff
@@ -23,34 +23,56 @@ const onKebabOptionClick = (option: KebabOption) => {
...
+) =>
+  connectorDropMenuCallback
+    ? createMenuItems(
+        graphActions(element.getController().getElements()),
+        connectorDropMenuCallback,
+      )
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: You don't need to skip calling `kebabOptionsToMenu` as it'll be a noop if there's no sub menus.
Also you won't need to pass in `connectorDropMenuCallback` anymore.

**Comment by christianvogt** [comment]
The flickering needs a followup PR. It's probably due to the async review checks on the kebab items :(

**Comment by christianvogt** [comment]
@andrewballantyne this PR will fix the arrowhead issue: https://github.com/openshift/console/pull/4063

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx (hunk 20 above)
Comment (nemesis09) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: Done

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx (hunk 21 above)
Comment (nemesis09) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: Done

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 22 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: updated the code to enable use of `add-resources-menu-utils`

**Comment by nemesis09** [comment]
@christianvogt 
I have updated the PR with the changes. PTAL

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts (hunk 16 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/componentUtils.ts]: the new changes takes care of it

**Comment by nemesis09** [comment]
> image
> 
> @nemesis09 You have a key issue in the console...

fixed

**Comment by andrewballantyne** [comment]
Clicking a group white-screens. `TypeError: Cannot read property 'getData' of undefined`

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx (hunk 20 above)
Comment (nemesis09) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: this change causes `key` error

**Comment by nemesis09** [comment]
> Clicking a group white-screens. `TypeError: Cannot read property 'getData' of undefined`

Added checks to fix the issue. PTAL

**Comment by andrewballantyne** [comment]
Not sure if this is you or @karthikjeeyar but the action menu that appears in the application group sidebar throws an error:

> Warning: Failed prop type: Invalid prop `actions[3].accessReview` of type `boolean` supplied to `ConnectFunction`, expected `object`.

**Comment by nemesis09** [comment]
> Not sure if this is you or @karthikjeeyar but the action menu that appears in the application group sidebar throws an error:
> 
> > Warning: Failed prop type: Invalid prop `actions[3].accessReview` of type `boolean` supplied to `ConnectFunction`, expected `object`.

I saw the warning too.
but it does not appear consistently. I got it a few times.

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx [hunk 26]
```diff
@@ -23,8 +23,8 @@ const onKebabOptionClick = (option: KebabOption) => {
   }
 };
 
-const createMenuItems = (actions: KebabMenuOption[]) =>
-  actions.map((option) =>
+const createMenuItems = (actions: KebabMenuOption[]) => {
+  return actions.map((option) =>
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: unnecessary change

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx [hunk 27]
```diff
@@ -36,21 +36,26 @@ const createMenuItems = (actions: KebabMenuOption[]) =>
...
+export const groupContextMenu = (element: Node, connectorSource?: Node) => {
   const applicationData: TopologyApplicationObject = {
     id: element.getId(),
     name: element.getLabel(),
     resources: element.getChildren().map((node: GraphElement) => node.getData()),
   };
 
-  return createMenuItems(kebabOptionsToMenu(groupActions(applicationData)));
+  return connectorSource
+    ? createMenuItems(groupActions(applicationData, connectorSource))
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: This needs a `TODO` comment for cleanup or a fix now.
We shouldn't have to avoid calling `kebabOptionsToMenu`. Instead the actions provided should omit including the `path` when used in this scenario.

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx [hunk 28]
```diff
@@ -36,21 +36,26 @@ const createMenuItems = (actions: KebabMenuOption[]) =>
...
+    : createMenuItems(kebabOptionsToMenu(groupActions(applicationData)));
 };
 export const nodeContextMenu = (element: Node) =>
   createMenuItems(kebabOptionsToMenu(nodeActions(element.getData())));
 
-export const graphContextMenu = (element: Graph) =>
-  createMenuItems(kebabOptionsToMenu(graphActions(element.getController().getElements())));
+export const graphContextMenu = (element: Graph, connectorSource?: Node) =>
+  connectorSource
+    ? createMenuItems(graphActions(element.getController().getElements(), connectorSource))
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: Same comment as above.

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx [hunk 29]
```diff
@@ -27,9 +27,10 @@ type CreateConnectorWidgetProps = {
     element: Node,
     target: Node | Graph,
     event: DragEvent,
-    choice?: ConnectorChoice,
-  ) => ConnectorChoice[] | void | undefined | null;
+    choice?: ConnectorChoice | React.ReactElement,
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/withCreateConnector.tsx]: This is incorrect, leave it. We don't callback with a `ReactElement`.
It was a mistake in my pseudo code posted earlier attempting to describe a new solution.

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx (hunk 20 above)
Comment (christianvogt) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: Sorry I forgot to remove this comment of mine .

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx (hunk 26 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: cleaned up

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx (hunk 27 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: fixed

File: frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx (hunk 28 above)
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/topology/nodeContextMenu.tsx]: fixed

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx (hunk 29 above)
Comment (nemesis09) [reply frontend/packages/topology/src/behavior/withCreateConnector.tsx]: fixed

**Comment by andrewballantyne** [comment]
/hold

Sam is doing investigations into the e2e, they're not likely to pass atm.
#4065

**Comment by andrewballantyne** [comment]
@nemesis09 They are not going to pass right now. PR #4065 is attempting to fix them.
